core/camel-main/src/main/java/org/apache/camel/main/MainDurationEventNotifier.java

```diff
@@ -110,10 +121,19 @@ protected void doNotify(CamelEvent event) {
             if (complete) {
                 doneMessages.increment();
                 final int doneCount = doneMessages.intValue();
-                final boolean result = doneCount >= maxMessages;
+                final int inflight = camelContext.getInflightRepository().size();
+
+                boolean result = doneCount >= maxMessages;
+                if (result && !maxMessagesIgnoreInflightExchanges) {
+                    result = inflight == 0;
+                }
 
                 if (LOG.isTraceEnabled()) {
-                    LOG.trace("Duration max messages check {} >= {} -> {}", doneCount, maxMessages, result);
+                    if (!maxMessagesIgnoreInflightExchanges) {
+                        LOG.trace("Duration max messages check {}/{} >= {} -> {}", doneCount, inflight, maxMessages, result);
+                    } else {
+                        LOG.trace("Duration max messages check {} >= {} -> {}", doneCount, maxMessages, result);
+                    }
```

Comment:
I'm wondering if we could have this in a way that retains the old logic. If you are processing a large batch of messages using the main, this part gets called quite a lot. So, this part was adjusted along with the performance fixes done around 4.0.0 development. 

As it is now, it is forcing the code to hit a path that is costlier (i.e.; the inflight repository instance is volatile, then the call itself may or many not be inlined as the JIT may not be able to determine its type, etc). 

So, I think having 2 separate methods would be better here.

Reply:
would moving the inflight repository check inside the if enough ?
 something like:

```java 
int inflight = 0;
boolean result = doneCount >= maxMessages;

if (result && !maxMessagesIgnoreInflightExchanges) {
    inflight = camelContext.getInflightRepository().size();
    result = inflight == 0;
}
```

Reply:
I think that would do the trick